### PR TITLE
Add lazy zipWithIndex

### PR DESCRIPTION
**Pull request by sadikovi**
This PR is essentially copy of of `org.apache.spark.rdd.ZippedWithIndexRDD` with some minor modification of computing length of each partition only when `partitions` are called or action is invoked on RDD (which does compute partitions). 

Had to move/reimplement some functionality for reasons of keeping under `org.hammerlab.magic.rdd`. 

Usage:

``` shell
scala> import org.hammerlab.magic.rdd.LazyZippedWithIndexRDD._
scala> val rdd = sc.parallelize(0 until 10).filter(_ % 2 == 0).lazyZipWithIndex
scala> rdd.collect
res1: Array[(Int, Long)] = Array((0,0), (2,1), (4,2), (6,3), (8,4))

```

### CODE REVIEW

**Comment by sadikovi** [comment]
Fixes #3.

**Comment by sadikovi** [comment]
@ryan-williams please review. 

I am not sure how this code matches idea in the issue, but I hope it solves lazy evaluation problem. Also trying to come up with some sort of shuffled zipWithIndex that does not result in computing twice, but it is far from this PR.

File: src/main/scala/org/hammerlab/magic/rdd/LazyZippedWithIndexRDD.scala [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+}
+
+/**
+ * Slight modification of [[ZippedWithIndexRDD]] from Spark that does not run job when created,
+ * it only triggers job on action or computing partitions.
+ */
+class LazyZippedWithIndexRDD[T: ClassTag](private var rdd: RDD[T]) extends RDD[(T, Long)](rdd) {
+  private var startIndices: Array[Long] = null
+
+  val iteratorSize: Iterator[_] => Long = iter => {
```
Comment (ryan-williams) [inline src/main/scala/org/hammerlab/magic/rdd/LazyZippedWithIndexRDD.scala]: ah, I changed this in https://github.com/sadikovi/magic-rdds/pull/1 for #8 as well… any reason not to use `iter.size`?

File: src/main/scala/org/hammerlab/magic/rdd/LazyZippedWithIndexRDD.scala [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    val n = rdd.partitions.length
+    if (n == 0) {
+      Array[Long]()
+    } else if (n == 1) {
+      Array(0L)
+    } else {
+      rdd.context.runJob(
+        rdd,
+        iteratorSize,
+        0 until n - 1 // do not need to count the last partition
```
Comment (ryan-williams) [inline src/main/scala/org/hammerlab/magic/rdd/LazyZippedWithIndexRDD.scala:35]: clever!

File: src/test/scala/org/hammerlab/magic/rdd/LazyZippedWithIndexRDDTest.scala [hunk 3]
```diff
@@ -0,0 +1,13 @@
+package org.hammerlab.magic.rdd
+
+import org.hammerlab.magic.rdd.LazyZippedWithIndexRDD._
+import org.hammerlab.magic.util.SparkSuite
+
+class LazyZippedWithIndexRDDTest extends SparkSuite {
+  test("match with zipWithIndex") {
+    val rdd = sc.parallelize(0 until 10).filter(_ % 2 == 0)
+    val res1 = rdd.lazyZipWithIndex.collect
+    val res2 = rdd.zipWithIndex.collect
```
Comment (ryan-williams) [inline src/test/scala/org/hammerlab/magic/rdd/LazyZippedWithIndexRDDTest.scala]: would also be nice to have a `TestSparkListener` check here; let me send a separate PR with that code, though that doesn't need to block merging this PR if you respond to my other comment above

**Comment by ryan-williams** [comment]
This looks great, thanks @sadikovi. One comment, about `Iterator.size`

File: src/main/scala/org/hammerlab/magic/rdd/LazyZippedWithIndexRDD.scala (hunk 1 above)
Comment (sadikovi) [inline src/main/scala/org/hammerlab/magic/rdd/LazyZippedWithIndexRDD.scala]: Iterator uses for loop to compute size, which, people were saying, was slightly slower, most likely will not matter at all.

Thanks, will update accordingly.

File: src/test/scala/org/hammerlab/magic/rdd/LazyZippedWithIndexRDDTest.scala (hunk 3 above)
Comment (sadikovi) [inline src/test/scala/org/hammerlab/magic/rdd/LazyZippedWithIndexRDDTest.scala]: I will change this test to use test listener, thanks.

**Comment by sadikovi** [comment]
Included minor update for test listener in this PR, though I was thinking about creating separate PR, but instead updated both with the same change by cherry-picking commit. I apologize, if this caused any problems with commit history for repository.

**Comment by sadikovi** [comment]
@ryan-williams addressed your comments, thanks for reviewing PR!

File: src/main/scala/org/hammerlab/magic/test/listener/TestSparkListener.scala [hunk 4]
```diff
@@ -158,7 +158,12 @@ object TestSparkListener {
 
   var instance: TestSparkListener = _
 
-  def apply(): TestSparkListener = instance
+  def apply(): TestSparkListener = {
+    if (instance == null) {
```
Comment (sadikovi) [inline src/main/scala/org/hammerlab/magic/test/listener/TestSparkListener.scala:157]: This is done for cases when listener calls `clear()`, but application is not initialised yet. Similar change is made in another PR, will resolve conflicts once one of them is merged.

**Comment by sadikovi** [comment]
Same update as for cached count, so I guess it is better to discuss it in that PR.

**Comment by ryan-williams** [comment]
https://github.com/sadikovi/magic-rdds/pull/3 has a couple tweaks to this then it is good to merge in!
